### PR TITLE
Added autodiscover method with generation of new permissions

### PR DESCRIPTION
**Pull request by pivancy**
TODO: Create logic with exclusion and removal of unneeded permissions.

### CODE REVIEW

File: view_acl/views.py [hunk 1]
```diff
@@ -1,2 +1,37 @@
...
+def generate_permissions(urlpatterns, permissions=[]):
+    """ Generate names for permissions. """
+    for pattern in urlpatterns:
+        if isinstance(pattern, urlresolvers.RegexURLPattern):
+            permissions.append(_generate_perm_name(pattern.callback))
+        elif isinstance(pattern, urlresolvers.RegexURLResolver):
+            generate_permissions(pattern.url_patterns, permissions)
+    return permissions
+
+def _generate_perm_name(callback):
```
Comment (ntodosiychuk) [inline view_acl/views.py]: Please move func to utils fie

File: view_acl/views.py [hunk 2]
```diff
@@ -1,2 +1,37 @@
-def autodiscovery():
-    pass
\ No newline at end of file
+from django.conf import settings
```
Comment (ntodosiychuk) [inline view_acl/views.py]: Seems this code should be moved to init or other file.
It isn't django view.

File: view_acl/utils.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        elif isinstance(pattern, urlresolvers.RegexURLResolver):
+            generate_permissions(pattern.url_patterns, permissions)
+    return permissions
+
+def _generate_perm_name(callback):
+    """ Generate permission name from callback. """
+    return '.'.join((callback and callback.__module__ or '',
+                     callback and callback.__name__ or ''))
+
+def create_permissions(codename, name):
```
Comment (pzdeb) [inline view_acl/utils.py]: this function creates only one permissions, the name must be singular

File: view_acl/__init__.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+def autodiscover():
+    """ Autodiscover for urls.py """
+    urls = getattr(settings, URLCONF, set())
+    resolver = urlresolvers.get_resolver(urls)
+    urlpatterns = resolver.url_patterns
+    permissions = generate_permissions(urlpatterns)
+    exclusion_perms = getattr(settings, 'EXCLUSION_LIST', [])
+    permissions = exclude_perms(exclusion_perms, permissions)
+    for perm in permissions:
```
Comment (pzdeb) [inline view_acl/__init__.py]: This method must also refresh all available permissions and remove unused ones

File: view_acl/__init__.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+URLCONF = 'ROOT_URLCONF'
+
+def autodiscover():
+    """ Autodiscover for urls.py """
+    urls = getattr(settings, URLCONF, set())
+    resolver = urlresolvers.get_resolver(urls)
+    urlpatterns = resolver.url_patterns
+    permissions = generate_permissions(urlpatterns)
+    exclusion_perms = getattr(settings, 'EXCLUSION_LIST', [])
```
Comment (pzdeb) [inline view_acl/__init__.py]: Default views and applications like Admin and auth must be in exclusion list by default

File: view_acl/utils.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+def generate_permissions(urlpatterns, permissions=[]):
+    """ Generate names for permissions. """
+    for pattern in urlpatterns:
+        if isinstance(pattern, urlresolvers.RegexURLPattern):
+            permissions.append(_generate_perm_name(pattern.callback))
+        elif isinstance(pattern, urlresolvers.RegexURLResolver):
+            generate_permissions(pattern.url_patterns, permissions)
+    return permissions
+
+def _generate_perm_name(callback):
```
Comment (pzdeb) [inline view_acl/utils.py]: please remove this leading dash and rename the simple parameter to func

File: view_acl/__init__.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+from django.conf import settings
+from django.core import urlresolvers
+
+from .utils import create_permissions, exclude_perms, generate_permissions
+
+URLCONF = 'ROOT_URLCONF'
+
+def autodiscover():
```
Comment (pzdeb) [inline view_acl/__init__.py]: Logic of permission creation must me smarter enough to handle existing and new permissions. Just get_or_create won't work here.
Proposed algorithm is following:
1. Get list of all `available_views_permissions`
2. Generate all `active_view_permissions` for active urls and views
3. Add new permissions from `active_view_permissions` that do not exist in `available_views_permissions`
4. Remove permissions from `available_views_permissions` that don't exist in `active_view_permissions`

**Comment by pivancy** [comment]
@pzdeb, @ntodosiychuk I've added removal logic for deprecated permissions + changed some names for functions,related to comments

**Comment by ntodosiychuk** [comment]
+
I will take your changes and continue work on new functionality.

**Comment by ntodosiychuk** [comment]
@pivancy 
I've merged your changes into my branch.
And I will close this PR.
Your changes are here
https://github.com/Perfectial/django-view-acl/pull/4

**Pull request closed without merging**